Wrap LLVMAttributeRef API

kyouko-taiga opened the pull request:
**Disclaimer:** I'm quite a newbie with LLVM

This is my attempt at wrapping the LLVM 4.0.0+ `LLVMAttributeRef` API (as mentioned in #10).

I extracted the list of parameters from [LLVM's docs](http://llvm.org/docs/LangRef.html#function-attributes). I didn't make any unit tests so far, and haven't tested all parameters.

Due to my limited knowledge of LLVM, I'm unsure of the expected result of calling `LLVMAddAttributeAtIndex` on parameters whose arguments aren't a single integer value (e.g. `allocsize`), nor of the difference between attributes specified with or without quote in the documentation (e.g. `"patchable-function"`). Some review from a more experienced developer would be greatly appreciated.

I tried to follow the coding convention I could guess from the existing code to the best of my ability, but don't hesitate to comment on that as well.

kyouko-taiga committed:
Wrap LLVMAttributeRef API

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,262 @@
+  public func addAttribute(_ attr: FunctionAttribute, value: UInt64 = 0) {
+    let ctx = LLVMGetModuleContext(LLVMGetGlobalParent(llvm))
+    let attrRef = LLVMCreateEnumAttribute(ctx, attr.kindID, value)
+    LLVMAddAttributeAtIndex(llvm, 0, attrRef)
```
This is not correct.  This is the index for the return value of the function.

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,262 @@
+  /// - parameterIndex: The index of the parameter to which add the
+  ///   attribute, starting from 0.
+  public func addParameterAttribute(_ attr: ParameterAttribute, value: UInt64 = 0,
+                                    for parameterIndex: LLVMAttributeIndex) {
```
We have a policy of never exposing LLVM's types to the user.  I have an idea of how we can lift this into an enum in one case.

CodaFi replied:
Please use `Int` here.

CodaFi commented:
I think we can kill two birds with one stone here by lifting `AttributeList`'s sorta ad-hoc indexing model to an enum:

```swift
enum AttributeIndex: UInt32 {
  case returnIndex // 0
  case functionIndex // ~0
  case argumentIndex(UInt)
}
```

The `RawRepresentable` conformance will have to be done by hand.

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,262 @@
+#endif
+
+/// Enumerates the attributes of LLVM functions.
+public enum FunctionAttribute: String {
```
Thank you for documenting this patch.

CodaFi commented:
Do you know if `lld` or the AttributeSet does bounds checking for the parameter indices?  If not we should put in a precondition.

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,262 @@
+  /// This attribute indicates that the function may only access memory that is
+  /// either not accessible by the module being compiled, or is pointed to by
+  /// its pointer arguments.
+  case inaccessiblemem_or_argmemonly
```
All the `snake_case` should be `camelCase`.

CodaFi commented:
Oh, it'd also be great if you could add a test.  It can be as simple as a loop that adds All The Attributes to a dummy function and its parameters and checks that they got registered.

kyouko-taiga committed:
Lift LLVMAttributeIndex with a raw representable enum

kyouko-taiga committed:
Rewrite snake_case to camelCase

kyouko-taiga committed:
Test `addAttribute`

kyouko-taiga committed:
Distinguish between enum and string attributes

kyouko-taiga committed:
Test set/unset attributes

kyouko-taiga commented:
Thanks for the feedbacks!

A few comments on the changes I made:

According to [LLVM docs](https://llvm.org/docs/HowToUseAttributes.html), I understand there are several kinds of attributes in LLVM's new scheme. If I understand the docs correctly, this is also where the distinction between attributes specified with or without quotes comes in. The formers describe target-independent attributes, while the latters describe dependent ones. Therefore I took the liberty to create two types to distinguish between those two kind of attributes. I wrapped all `LLVM*StringAttribute*` functions, which I believe were part of LLVMSwift missing functionalities, as mentioned in #51 .

I implemented a method to retrieve the attributes from a function. I made the choice to return an array rather than a set, so as to avoid having to conform to `Hashable` (which seemed a bit overkill to me). Unfortunately, it seems like there's no wrapper around C++'s [`Attribute::getKindAsString()`](http://llvm.org/doxygen/classllvm_1_1Attribute.html#a3ad8f83f8d6165314fe8f173645dd458) so I can't provide a proper implementation of `EnumAttribute.name`.

I wrote some tests for all these shenanigans. Those revealed that there seems to be a problem with the removal of valued enum attributes. I tried to understand what's causing this bug within LLVM's source, but I have to admit to find the whole thing quite puzzling. Indeed, removing an integer attribute seems to always fail, because [`AttributeList::removeAttribute()`](http://llvm.org/doxygen/classllvm_1_1AttributeSet.html#a1247d3316a7816eaee459556671caab5) attempts to add the attribute about to be removed without preserving its value. Maybe an issue with my version of LLVM ...

kyouko-taiga commented:
I confirm the bug about valued enum attributes was caused from LLVM: [[Attributes] Fix crash when attempting to remove alignment from an attribute list/set](https://github.com/llvm-mirror/llvm/commit/9bc0b1080f195636fed019bce979aa72892d6c69#diff-19a9dba7179bdf5da40ec95b8bfefe82).

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,374 @@
+
+  /// The name of the attribute's kind.
+  public var name: String {
+    return ""
```
If we can't provide this then we should omit it.

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,374 @@
+}
+
+/// An LLVM attribute.
+public protocol Attribute {
```
Can you document this and its members a little more

CodaFi commented on Sources/LLVM/Function+Attributes.swift:
```diff
@@ -0,0 +1,374 @@
+      let cstring = LLVMGetStringAttributeKind(stringAttr.llvm, &length)
+      LLVMRemoveStringAttributeAtIndex(llvm, index.rawValue, cstring, length)
+    default:
+      fatalError()
```
Please add an explanation string to this fatal error

CodaFi commented:
Down to nits.  This is looking quite good.

kyouko-taiga committed:
Remove `name` property requirements from `Attribute`

kyouko-taiga committed:
Provide an explanation for failure to cast attribute

kyouko-taiga committed:
Add some documentation

kyouko-taiga commented:
What would be the best thing to do about the issue with `removeAttribute(:from:)` on valued enum attributes? The issue seems to be [fixed in LLVM 7.0.0svn](http://llvm.org/doxygen/classllvm_1_1AttributeSet.html#a1247d3316a7816eaee459556671caab5) but if I remove the guard in [`IRAttributesSpec.swift:138`](https://github.com/kyouko-taiga/LLVMSwift/blob/71095452fdfa5358bf19b686d3a465b8760da2d0/Tests/LLVMTests/IRAttributesSpec.swift#L138) the tests won't pass with previous versions. I don't know what's LLVMSwift's stance about the version of LLVM it is linked to.

CodaFi commented:
If we can't safely provide this API then we shouldn't provide it altogether.  We can hold out until 7.0 and leave behind a note.  I'd much rather that than crashing deep in LLVM.

CodaFi commented:
What I would recommend is we bake the guard into the API and fatal error with an explanation as to why we cannot safely remove the attribute yet.

kyouko-taiga commented:
Is there any way we can know what version of LLVM we're linked to?

CodaFi commented:
`LLVM_VERSION_MAJOR/MINOR/PATCH` but we tie ourselves to the latest version as it is announced and do not do backwards compatibility.

kyouko-taiga commented:
Alright, I'd make the guard only for the attributes that may file, namely `align`, `alignstack`, `allocsize`, `dereferenceable` and `dereferenceable_or_null`.

kyouko-taiga committed:
Add an error message when trying to remove valued enum attributes

The current version of LLVM crashes when attempting to remove valued
enum attributes. The bug was discovered and fixed in LLVM 7 (see
https://github.com/llvm-mirror/llvm/commit/9bc0b10). The guard placed
in `removeAttribute` should be removed once this version is released.

CodaFi commented:
Excellent.  Thank you kindly.

⛵️
